server/src/main/java/org/opensearch/cluster/decommission/DecommissionHelper.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.decommission;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.ActionListener;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.ClusterStateObserver;
+import org.opensearch.cluster.ClusterStateTaskConfig;
+import org.opensearch.cluster.ClusterStateTaskListener;
+import org.opensearch.cluster.ack.ClusterStateUpdateResponse;
+import org.opensearch.cluster.coordination.NodeRemovalClusterStateTaskExecutor;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.cluster.routing.allocation.AllocationService;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.Priority;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+
+/**
+ * Helper executor class to remove list of nodes from the cluster
+ *
+ * @opensearch.internal
+ */
+
+public class DecommissionHelper {
+
+    private static final Logger logger = LogManager.getLogger(DecommissionHelper.class);
+
+    private final NodeRemovalClusterStateTaskExecutor nodeRemovalExecutor;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+
+    DecommissionHelper(
+        ClusterService clusterService,
+        AllocationService allocationService,
+        ThreadPool threadPool
+    ) {
+        this.clusterService = clusterService;
+        this.nodeRemovalExecutor = new NodeRemovalClusterStateTaskExecutor(allocationService, logger);
+        this.threadPool = threadPool;
+    }
+
+    public void handleNodesDecommissionRequest(
```

Comment:
It doesn't wait for shards to move away from the node. Is it ok for it to cause cluster to go red? shouldn't it fail the shards first, so that they move out.

Reply:
We don't want the shards to move out, as it might trigger a storm of shard recoveries impacting latencies due to additional CPU and network consumption. We will let shard stay in UNASSIGNED state after decommissioning the zone. 

We will ensure that decommissioning is graceful and no inflight requests are impacted but also expects that the users has enough shard copies on each zone before decommissioning one. The system allows only one zone to be decommissioned at a time and also when the awareness and forced awareness is enabled. Users can use `AwarenessReplicaBalance` to enforce shard copies be multiple of total awareness attributes